src/AI/Muriel.java

```diff
@@ -405,8 +405,30 @@ public GameAction getNextAction(GameMap gameMap)
           UnitMatchupAndMetaInfo umami = getUnitMatchupInfo(unit, target);
           if( umami.costEffectivenessRatio < COST_EFFECTIVENESS_MIN ) continue;
 
+          // Find locations that would be dangerous for us so we can avoid sauntering into enemy fire.
+          HashSet<XYCoord> noGoZone = new HashSet<XYCoord>();
+          final int MAX_RELEVANT_DISTANCE = 27; // ~3x the move distance of the fastest units. Up for tweaking.
+          for(int j = 0; j < enemyLocations.size(); ++j)
+          {
+            XYCoord threatCoord = enemyLocations.get(j);
+            Unit threat = gameMap.getLocation(threatCoord).getResident();
+            XYCoord unitCoord = new XYCoord(unit.x, unit.y);
+            if( Utils.findManhattanDistance(unitCoord, threatCoord) <= MAX_RELEVANT_DISTANCE )
+            {
+              // If we, in the enemy's place, would attack `unit` with `threat`, then we should not let them attack us.
+              UnitMatchupAndMetaInfo invUmami = getUnitMatchupInfo(threat, unit);
+              if( invUmami.costEffectivenessRatio > COST_EFFECTIVENESS_MIN && threat.getHP() >= 6 )
+              {
+                // Add coordinates that `threat` could target to our "no-go" list.
+                Map<XYCoord, Double> threatMap = AIUtils.findThreatPower(gameMap, threat, unit.model);
+                noGoZone.addAll(threatMap.keySet()); // Ignore the valueMap of the return; we have already decided `threat` is dangerous.
+              }
+            }
+            else break; // Don't bother considering far-away baddies for our no-go zone.
```

Comment:
I'd change these to `if x break;` and `if y continue;`, since I find that construct more readable than enclosing stuff in braces.

Reply:
Not sure what you mean. This line has no braces, and there is no `continue` here. Do you mean to invert the conditional above and do
```java
if( !( Utils.findManhattanDistance(unitCoord, threatCoord) <= MAX_RELEVANT_DISTANCE ) )
  break;
else
{
  // Do the logic
}
```

This way makes sense to me. If conditions are met, do all this logic, otherwise just skip it.

Reply:
More like:
```java
if( unitCoord.getDistance(threatCoord) >= MAX_RELEVANT_DISTANCE )
  break; // No more relevant threats
if( !threat.canAttack(unit.model) && shouldAttack(threat, unit, gameMap) )
  continue; // This threat isn't one
// Do logic
```
Either way makes sense, but fewer syntactically relevant lines (assuming you aren't arbitrarily grouping things) makes the code more readable IMO.
In this case, you no longer have an `if...else` to keep track of and two fewer scopes to keep track of.

Reply:
Ah, that makes sense. I'll do things like that for longer functions sometimes, but for this I didn't bother. Maybe I'll change it if I have to go back in and add/refactor stuff later.